src/station_cmd.cpp

```diff
@@ -3779,44 +3779,22 @@ CommandCost CmdRenameStation(TileIndex tile, DoCommandFlag flags, uint32 p1, uin
  */
 void FindStationsAroundTiles(const TileArea &location, StationList *stations)
 {
-	/* area to search = producer plus station catchment radius */
-	uint max_rad = (_settings_game.station.modified_catchment ? MAX_CATCHMENT : CA_UNMODIFIED);
-
 	uint x = TileX(location.tile);
 	uint y = TileY(location.tile);
 
-	uint min_x = (x > max_rad) ? x - max_rad : 0;
-	uint max_x = x + location.w + max_rad;
-	uint min_y = (y > max_rad) ? y - max_rad : 0;
-	uint max_y = y + location.h + max_rad;
-
-	if (min_x == 0 && _settings_game.construction.freeform_edges) min_x = 1;
-	if (min_y == 0 && _settings_game.construction.freeform_edges) min_y = 1;
-	if (max_x >= MapSizeX()) max_x = MapSizeX() - 1;
-	if (max_y >= MapSizeY()) max_y = MapSizeY() - 1;
-
-	for (uint cy = min_y; cy < max_y; cy++) {
-		for (uint cx = min_x; cx < max_x; cx++) {
-			TileIndex cur_tile = TileXY(cx, cy);
-			if (!IsTileType(cur_tile, MP_STATION)) continue;
-
-			Station *st = Station::GetByTile(cur_tile);
-			/* st can be NULL in case of waypoints */
-			if (st == NULL) continue;
+	/* Don't like this ALL stations :/ */
+	Station *st;
+	FOR_ALL_STATIONS(st) {
+		bool included = false;
 
-			if (_settings_game.station.modified_catchment) {
-				int rad = st->GetCatchmentRadius();
-				int rad_x = cx - x;
-				int rad_y = cy - y;
+		for (uint cy = y; cy < y + location.h && !included; cy++) {
+			for (uint cx = x; cx < x + location.w; cx++) {
+				if (!st->TileIsInCatchment(TileXY(cx, cy))) continue;
```

Comment:
Could it improve performance to skip these loops based on a simple distance check for `location` near station sign?

Reply:
I don't think so. Very often, location.h and location.w are just 1 anyway. And we can't just do a distance check because catchment distance depends on the type of station tile.

Reply:
A distance check could be worst-case, the main idea would be to filter out stations it definitely could not be.

Reply:
Ah, yes, it reduces the time from 39ms down ot... 5ms. Pretty good. I wonder if the nearby stations caching is worth the complexity. Probably.

Reply:
5ms down to 1.5ms is probably worth having, yeah.